rcl/test/rcl/test_security.cpp

```diff
@@ -0,0 +1,221 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+
+#include <algorithm>
+#include <string>
+
+#include "rcl/security.h"
+#include "rcl/error_handling.h"
+
+#include "rcutils/filesystem.h"
+
+#include "rmw/error_handling.h"
+
+#include "osrf_testing_tools_cpp/scope_exit.hpp"
+
+
+#define TEST_SECURITY_DIRECTORY_RESOURCES_DIR_NAME "/test_security_directory"
+#define TEST_SECURITY_CONTEXT "dummy_security_context"
+#define TEST_SECURITY_CONTEXT_ABSOLUTE "/" TEST_SECURITY_CONTEXT
+
+#ifndef _WIN32
+# define PATH_SEPARATOR "/"
+#else
+# define PATH_SEPARATOR "\\"
+#endif
+
+char g_envstring[512] = {0};
+
+static int putenv_wrapper(const char * env_var)
+{
+#ifdef _WIN32
+  return _putenv(env_var);
+#else
+  return putenv(reinterpret_cast<char *>(const_cast<char *>(env_var)));
+#endif
+}
+
+static int unsetenv_wrapper(const char * var_name)
+{
+#ifdef _WIN32
+  // On windows, putenv("VAR=") deletes VAR from environment
+  std::string var(var_name);
+  var += "=";
+  return _putenv(var.c_str());
+#else
+  return unsetenv(var_name);
+#endif
+}
+
+class TestGetSecureRoot : public ::testing::Test
+{
+protected:
+  void SetUp() final
+  {
+    // Reset rcutil error global state in case a previously
+    // running test has failed.
+    rcl_reset_error();
+
+    // Always make sure the variable we set is unset at the beginning of a test
+    unsetenv_wrapper(ROS_SECURITY_ROOT_DIRECTORY_VAR_NAME);
+    unsetenv_wrapper(ROS_SECURITY_DIRECTORY_OVERRIDE);
+    unsetenv_wrapper(ROS_SECURITY_STRATEGY_VAR_NAME);
+    unsetenv_wrapper(ROS_SECURITY_ENABLE_VAR_NAME);
+    allocator = rcl_get_default_allocator();
+    root_path = nullptr;
+    secure_root = nullptr;
+    base_lookup_dir_fqn = nullptr;
+  }
+
+  void TearDown() final
+  {
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      allocator.deallocate(root_path, allocator.state);
+    });
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      allocator.deallocate(secure_root, allocator.state);
+    });
+    OSRF_TESTING_TOOLS_CPP_SCOPE_EXIT(
+    {
+      allocator.deallocate(base_lookup_dir_fqn, allocator.state);
+    });
+  }
+
+  void set_base_lookup_dir_fqn(const char * resource_dir, const char * resource_dir_name)
+  {
+    base_lookup_dir_fqn = rcutils_join_path(
+      resource_dir, resource_dir_name, allocator);
+    std::string putenv_input = ROS_SECURITY_ROOT_DIRECTORY_VAR_NAME "=";
+    putenv_input += base_lookup_dir_fqn;
+    memcpy(
+      g_envstring, putenv_input.c_str(),
+      std::min(putenv_input.length(), sizeof(g_envstring) - 1));
+    putenv_wrapper(g_envstring);
+  }
+
+  rcl_allocator_t allocator;
+  char * root_path;
+  char * secure_root;
+  char * base_lookup_dir_fqn;
+};
+
+TEST_F(TestGetSecureRoot, failureScenarios) {
+  ASSERT_EQ(
```

Comment:
@ivanpauno here and elsewhere, do we have to `assert` or would an `expect` perform equally well? Particularly in those tests that currently have a single assertion.

Reply:
See https://github.com/ros2/rcl/pull/515/commits/e3d5ebdf6094274f7f74d752ecce3fbadec84150.